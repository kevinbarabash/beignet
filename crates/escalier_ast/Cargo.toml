[package]
name = "escalier_ast"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
swc_atoms = "0.5.6"
swc_ecma_ast = "0.107.0"
swc_common = { version = "0.31.16", features = ["sourcemap"] }
generational-arena = "0.2.8"
