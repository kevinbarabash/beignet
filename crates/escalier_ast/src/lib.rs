pub mod block;
pub mod class;
pub mod expr;
pub mod func_param;
pub mod identifier;
pub mod jsx;
pub mod literal;
pub mod pattern;
pub mod script;
pub mod span;
pub mod stmt;
pub mod type_ann;
pub mod type_param;
pub mod visitor;

pub use block::*;
pub use class::*;
pub use expr::*;
pub use func_param::*;
pub use identifier::*;
pub use jsx::*;
pub use literal::*;
pub use pattern::*;
pub use script::*;
pub use span::*;
pub use stmt::*;
pub use type_ann::*;
pub use type_param::*;
pub use visitor::*;
