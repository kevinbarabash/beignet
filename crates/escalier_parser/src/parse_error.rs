use std::fmt;

// TODO: turn this into an enum so that we can report different kinds of parse
// errors
#[derive(Debug, PartialEq, Eq)]
pub struct ParseError {
    pub message: String,
}

impl fmt::Display for ParseError {
    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
        let Self { message } = self;
        write!(fmt, "ParseError: {message}")
    }
}

impl From<&str> for ParseError {
    fn from(message: &str) -> Self {
        ParseError {
            message: message.to_owned(),
        }
    }
}

impl From<String> for ParseError {
    fn from(message: String) -> Self {
        ParseError { message }
    }
}
