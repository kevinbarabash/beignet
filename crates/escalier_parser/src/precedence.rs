use lazy_static::lazy_static;
use std::collections::hash_map::HashMap;

pub type Precedence = u32;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Associativity {
    Left,
    Right,
    Neither, // indicates a unary operator
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum Operator {
    // 18
    // Grouping, - parsed as an atom

    // 17
    MemberAccess,
    OptionalChaining,
    ComputedMemberAccess,
    NewWithArgumentList,
    FunctionCall,
    TemplateLiteral,

    // 16
    // NewWithoutArgList,

    // 15
    // PostfixIncrement,
    // PostfixDecrement,

    // 14
    LogicalNot,
    // BitwiseNot,
    UnaryPlus,
    UnaryMinus,
    // PrefixIncrement,
    // PrefixDecrement,
    Typeof,
    Void,
    Delete,
    Await,
    Throw,

    // 13
    Exponentiation,

    // 12
    Multiplication,
    Division,
    Remainder,

    // 11
    Addition,
    Subtraction,

    // 10
    // BitwiseLeftShift,
    // BitwiseRightShift,
    // BitwiseUnsignedRightShift,

    // 9
    LessThan,
    LessThanOrEqual,
    GreaterThan,
    GreaterThanOrEqual,
    In,
    Instanceof,

    // 8
    Equals,    // always strict
    NotEquals, // always strict
    // StrictEquals,
    // StrictNotEquals,

    // 7
    // BitwiseAnd,

    // 6
    // BitwiseXor,

    // 5
    // BitwiseOr,

    // 4
    LogicalAnd,

    // 3
    LogicalOr,
    NullishCoalescing,

    // 2
    Assignment,
    Conditional,
    // Arrow, - superseded by `fn`
    Yield,
    YieldStar,
    Spread,
    // 1
    // CommaSequence,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct OpInfo {
    prec: Precedence,
    assoc: Associativity,
}

impl OpInfo {
    pub fn new_prefix(prec: Precedence) -> Self {
        Self {
            prec,
            assoc: Associativity::Neither,
        }
    }

    pub fn new_infix(prec: Precedence, assoc: Associativity) -> Self {
        Self { prec, assoc }
    }

    pub fn new_postfix(prec: Precedence) -> Self {
        Self {
            prec,
            assoc: Associativity::Neither,
        }
    }

    pub fn normalized_prec(&self) -> Precedence {
        self.prec * 10
    }

    pub fn infix_postfix_prec(&self) -> Precedence {
        let normalized = self.normalized_prec();
        match self.assoc {
            Associativity::Left => normalized,
            Associativity::Right => normalized - 1,
            Associativity::Neither => normalized + 1,
        }
    }
}

lazy_static! {
    pub static ref PRECEDENCE_TABLE: HashMap<Operator, OpInfo> = {
        let mut table: HashMap<Operator, OpInfo> = HashMap::new();

        table.insert(
            Operator::MemberAccess,
            OpInfo::new_infix(17, Associativity::Left),
        );
        table.insert(
            Operator::OptionalChaining,
            OpInfo::new_infix(17, Associativity::Left),
        );
        table.insert(Operator::ComputedMemberAccess, OpInfo::new_postfix(17));
        table.insert(Operator::NewWithArgumentList, OpInfo::new_postfix(17));
        table.insert(Operator::FunctionCall, OpInfo::new_postfix(17));
        table.insert(Operator::TemplateLiteral, OpInfo::new_postfix(17));

        table.insert(Operator::LogicalNot, OpInfo::new_prefix(14));
        table.insert(Operator::UnaryPlus, OpInfo::new_prefix(14));
        table.insert(Operator::UnaryMinus, OpInfo::new_prefix(14));
        table.insert(Operator::Typeof, OpInfo::new_prefix(14));
        table.insert(Operator::Void, OpInfo::new_prefix(14));
        table.insert(Operator::Delete, OpInfo::new_prefix(14));
        table.insert(Operator::Await, OpInfo::new_prefix(14));
        table.insert(Operator::Throw, OpInfo::new_prefix(14));

        table.insert(
            Operator::Exponentiation,
            OpInfo::new_infix(13, Associativity::Right),
        );

        table.insert(
            Operator::Multiplication,
            OpInfo::new_infix(12, Associativity::Left),
        );
        table.insert(
            Operator::Division,
            OpInfo::new_infix(12, Associativity::Left),
        );
        table.insert(
            Operator::Remainder,
            OpInfo::new_infix(12, Associativity::Left),
        );

        table.insert(
            Operator::Addition,
            OpInfo::new_infix(11, Associativity::Left),
        );
        table.insert(
            Operator::Subtraction,
            OpInfo::new_infix(11, Associativity::Left),
        );

        table.insert(
            Operator::LessThan,
            OpInfo::new_infix(9, Associativity::Left),
        );
        table.insert(
            Operator::LessThanOrEqual,
            OpInfo::new_infix(9, Associativity::Left),
        );
        table.insert(
            Operator::GreaterThan,
            OpInfo::new_infix(9, Associativity::Left),
        );
        table.insert(
            Operator::GreaterThanOrEqual,
            OpInfo::new_infix(9, Associativity::Left),
        );
        table.insert(Operator::In, OpInfo::new_infix(9, Associativity::Left));
        table.insert(
            Operator::Instanceof,
            OpInfo::new_infix(9, Associativity::Left),
        );

        table.insert(Operator::Equals, OpInfo::new_infix(8, Associativity::Left));
        table.insert(
            Operator::NotEquals,
            OpInfo::new_infix(8, Associativity::Left),
        );

        table.insert(
            Operator::LogicalAnd,
            OpInfo::new_infix(4, Associativity::Left),
        );

        table.insert(
            Operator::LogicalOr,
            OpInfo::new_infix(3, Associativity::Left),
        );
        table.insert(
            Operator::NullishCoalescing,
            OpInfo::new_infix(3, Associativity::Left),
        );

        table.insert(
            Operator::Assignment,
            OpInfo::new_infix(2, Associativity::Neither),
        );
        table.insert(
            Operator::Conditional,
            OpInfo::new_infix(2, Associativity::Left),
        );
        table.insert(Operator::Yield, OpInfo::new_infix(2, Associativity::Right));
        table.insert(
            Operator::YieldStar,
            OpInfo::new_infix(2, Associativity::Right),
        );
        table.insert(Operator::Spread, OpInfo::new_infix(2, Associativity::Right));

        table
    };
}
