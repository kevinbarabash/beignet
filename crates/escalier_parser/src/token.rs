use crate::expr::Expr;
use crate::jsx::{JSXElement, JSXFragment};
use crate::source_location::SourceLocation;

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum TokenKind {
    Identifier(String), // [a-zA-Z_][a-zA-Z0-9_]*

    // Literals
    BoolLit(bool),
    NumLit(String),
    StrLit(String),
    StrTemplateLit {
        parts: Vec<Token>, // This should only contain StrLit tokens
        exprs: Vec<Expr>,
    },
    Null,
    Undefined,
    JSXElement(JSXElement),
    JSXFragment(JSXFragment),

    // Types
    Number,
    Boolean,
    String,
    Symbol,

    // Keywords
    Let,
    Return,
    Fn,
    Await,
    Async,
    If,
    Else,
    Mut, // denotes a binding to a mutable reference
    Var, // denotes a re-assignable binding
    Match,
    Try,
    Catch,
    Finally,
    Do,

    // Binary Operators
    Plus,
    Minus,
    Times,
    Divide,
    Modulo,
    Equals,
    NotEquals,
    LessThan,
    LessThanOrEqual,
    GreaterThan,
    GreaterThanOrEqual,
    Or,
    And,

    // Assignment Operators
    Assign,
    PlusAssign,
    MinusAssign,
    TimesAssign,
    DivideAssign,
    ModuloAssign,

    // punctuation
    Semicolon,
    Comma,
    LeftParen,
    RightParen,
    LeftBrace,
    RightBrace,
    LeftBracket,
    RightBracket,
    Arrow,
    Underscore,
    Question,
    QuestionDot, // used for optional chaining
    Dot,
    DotDot,    // used for ranges
    DotDotDot, // used for rest/spread
    Colon,

    Eof,
}

#[derive(Debug, PartialEq, Eq, Clone)]

pub struct Token {
    pub kind: TokenKind,
    pub loc: SourceLocation,
}
