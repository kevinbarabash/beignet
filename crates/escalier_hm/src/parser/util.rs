use crate::ast::*;

use crate::parser::parse_error::ParseError;
use crate::parser::pattern::parse_pattern;
use crate::parser::type_ann::parse_type_ann;

pub fn text_for_node(node: &tree_sitter::Node, src: &str) -> Result<String, ParseError> {
    match src.get(node.byte_range()) {
        Some(text) => Ok(text.to_owned()),
        None => Err(ParseError::from("error in text_for_node")),
    }
}

pub fn parse_type_params_for_node(
    node: &tree_sitter::Node,
    src: &str,
) -> Result<Option<Vec<TypeParam>>, ParseError> {
    let type_params = match node.child_by_field_name("type_parameters") {
        Some(type_params) => {
            let mut cursor = type_params.walk();
            let type_params = type_params
                .named_children(&mut cursor)
                .into_iter()
                .map(|type_param| {
                    let name_node = type_param.child_by_field_name("name").unwrap();
                    let name = Ident {
                        loc: SourceLocation::from(&name_node),
                        span: name_node.byte_range(),
                        name: text_for_node(&name_node, src)?,
                    };

                    let constraint =
                        if let Some(constraint) = type_param.child_by_field_name("constraint") {
                            Some(Box::from(parse_type_ann(&constraint, src)?))
                        } else {
                            None
                        };

                    let default = if let Some(value) = type_param.child_by_field_name("value") {
                        Some(Box::from(parse_type_ann(&value, src)?))
                    } else {
                        None
                    };

                    Ok(TypeParam {
                        span: type_param.byte_range(),
                        name,
                        constraint,
                        default,
                    })
                })
                .collect::<Result<Vec<_>, ParseError>>()?;
            Some(type_params)
        }
        None => None,
    };

    Ok(type_params)
}

pub fn parse_formal_parameters(
    node: &tree_sitter::Node,
    src: &str,
) -> Result<Vec<EFnParam>, ParseError> {
    assert_eq!(node.kind(), "formal_parameters");

    let mut cursor = node.walk();
    node.named_children(&mut cursor)
        .into_iter()
        .map(|param| {
            let optional = match param.kind() {
                "required_parameter" => false,
                "optional_parameter" => true,
                kind => panic!("Unexpected param kind: {kind}"),
            };
            let pattern = param.child_by_field_name("pattern").unwrap();
            let type_ann = if let Some(type_ann) = param.child_by_field_name("type") {
                Some(parse_type_ann(&type_ann, src)?)
            } else {
                None
            };

            Ok(EFnParam {
                pat: parse_pattern(&pattern, src)?,
                type_ann,
                optional,
            })
        })
        .collect::<Result<Vec<_>, ParseError>>()
}
