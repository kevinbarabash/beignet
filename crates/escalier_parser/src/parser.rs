use std::iter::{Iterator, Peekable};

use crate::source_location::*;
use crate::token::{Token, TokenKind};

pub struct Parser {
    tokens: Peekable<Box<dyn Iterator<Item = Token>>>,
}

impl Parser {
    pub fn new(tokens: Box<dyn Iterator<Item = Token>>) -> Self {
        Self {
            tokens: tokens.peekable(),
        }
    }

    pub fn next(&mut self) -> Token {
        self.tokens.next().unwrap_or(Token {
            kind: TokenKind::Eof,
            loc: SourceLocation {
                start: Position { line: 0, column: 0 },
                end: Position { line: 0, column: 0 },
            },
        })
    }

    pub fn peek(&mut self) -> Token {
        self.tokens.peek().cloned().unwrap_or(Token {
            kind: TokenKind::Eof,
            loc: SourceLocation {
                start: Position { line: 0, column: 0 },
                end: Position { line: 0, column: 0 },
            },
        })
    }
}
