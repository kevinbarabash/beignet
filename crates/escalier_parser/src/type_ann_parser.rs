use escalier_ast::*;

use crate::parse_error::ParseError;
use crate::parser::*;
use crate::precedence::Associativity;
use crate::token::*;

fn get_infix_precedence(op: &Token) -> Option<(u8, Associativity)> {
    match &op.kind {
        TokenKind::Ampersand => Some((11, Associativity::Left)),
        TokenKind::Pipe => Some((10, Associativity::Left)),
        _ => None,
    }
}

fn get_postfix_precedence(op: &Token) -> Option<(u8, Associativity)> {
    match &op.kind {
        TokenKind::LeftBracket => Some((12, Associativity::NotApplicable)),
        _ => None,
    }
}

impl<'a> Parser<'a> {
    fn parse_type_ann_atom(&mut self) -> Result<TypeAnn, ParseError> {
        let mut span = self.peek().unwrap_or(&EOF).span;
        let kind = match self.peek().unwrap_or(&EOF).kind.clone() {
            TokenKind::BoolLit(value) => {
                self.next();
                TypeAnnKind::BoolLit(value)
            }
            TokenKind::Boolean => {
                self.next();
                TypeAnnKind::Boolean
            }
            TokenKind::NumLit(value) => {
                self.next();
                TypeAnnKind::NumLit(value)
            }
            TokenKind::Number => {
                self.next();
                TypeAnnKind::Number
            }
            TokenKind::StrLit(value) => {
                self.next();
                TypeAnnKind::StrLit(value)
            }
            TokenKind::String => {
                self.next();
                TypeAnnKind::String
            }
            TokenKind::Symbol => {
                self.next();
                TypeAnnKind::Symbol
            }
            TokenKind::Null => {
                self.next();
                TypeAnnKind::Null
            }
            TokenKind::Undefined => {
                self.next();
                TypeAnnKind::Undefined
            }
            TokenKind::Unknown => {
                self.next();
                TypeAnnKind::Unknown
            }
            TokenKind::Never => {
                self.next();
                TypeAnnKind::Never
            }
            TokenKind::LeftBrace => {
                self.next(); // consumes '{'
                let mut props: Vec<ObjectProp> = vec![];

                while self.peek().unwrap_or(&EOF).kind != TokenKind::RightBrace {
                    match self.next().unwrap_or(EOF.clone()).kind {
                        TokenKind::Identifier(name) => {
                            let optional =
                                if self.peek().unwrap_or(&EOF).kind == TokenKind::Question {
                                    self.next().unwrap_or(EOF.clone());
                                    true
                                } else {
                                    false
                                };
                            assert_eq!(self.next().unwrap_or(EOF.clone()).kind, TokenKind::Colon);
                            let type_ann = self.parse_type_ann()?;

                            props.push(ObjectProp::Prop(type_ann::Prop {
                                name,
                                optional,
                                mutable: false, // TODO
                                type_ann: Box::new(type_ann),
                                span: Span { start: 0, end: 0 }, // TODO
                            }));
                        }
                        TokenKind::LeftBracket => {
                            let name = match self.next().unwrap_or(EOF.clone()).kind {
                                TokenKind::Identifier(name) => name,
                                _ => {
                                    return Err(ParseError {
                                        message: "expected identifier".to_string(),
                                    })
                                }
                            };
                            assert_eq!(self.next().unwrap_or(EOF.clone()).kind, TokenKind::Colon);
                            let type_ann = self.parse_type_ann()?;
                            assert_eq!(
                                self.next().unwrap_or(EOF.clone()).kind,
                                TokenKind::RightBracket
                            );

                            let key = IndexerKey {
                                name,
                                type_ann: Box::new(type_ann),
                            };

                            assert_eq!(self.next().unwrap_or(EOF.clone()).kind, TokenKind::Colon);
                            let type_ann = self.parse_type_ann()?;

                            props.push(ObjectProp::Indexer(Indexer {
                                key,
                                mutable: false, // TODO
                                type_ann: Box::new(type_ann),
                                span: Span { start: 0, end: 0 }, // TODO
                            }));
                        }
                        TokenKind::Fn => {
                            match self.peek().unwrap_or(&EOF).kind.clone() {
                                TokenKind::Identifier(name) => {
                                    self.next(); // consume identifier

                                    let type_params = self.maybe_parse_type_params()?;
                                    let params = self.parse_params()?;
                                    assert_eq!(
                                        self.next().unwrap_or(EOF.clone()).kind,
                                        TokenKind::Colon
                                    );
                                    let ret = self.parse_type_ann()?;

                                    props.push(ObjectProp::Method(ObjMethod {
                                        name,
                                        type_params,
                                        params,
                                        ret: Box::new(ret),
                                        span: Span { start: 0, end: 0 }, // TODO
                                    }));
                                }
                                TokenKind::LeftParen => {
                                    let type_params = self.maybe_parse_type_params()?;
                                    let params = self.parse_params()?;
                                    assert_eq!(
                                        self.next().unwrap_or(EOF.clone()).kind,
                                        TokenKind::Colon
                                    );
                                    let ret = self.parse_type_ann()?;

                                    props.push(ObjectProp::Call(ObjCallable {
                                        type_params,
                                        params,
                                        ret: Box::new(ret),
                                        span: Span { start: 0, end: 0 }, // TODO
                                    }));
                                }
                                _ => {
                                    return Err(ParseError {
                                        message: "expected identifier or left paren".to_string(),
                                    })
                                }
                            }
                        }
                        TokenKind::Get => {
                            let name = match self.next().unwrap_or(EOF.clone()).kind {
                                TokenKind::Identifier(name) => name,
                                _ => {
                                    return Err(ParseError {
                                        message: "expected identifier".to_string(),
                                    })
                                }
                            };

                            let params = self.parse_params()?;
                            assert_eq!(self.next().unwrap_or(EOF.clone()).kind, TokenKind::Colon);
                            let ret = self.parse_type_ann()?;

                            props.push(ObjectProp::Getter(ObjGetter {
                                name,
                                ret: Box::new(ret),
                                params,
                                span: Span { start: 0, end: 0 }, // TODO
                            }));
                        }
                        TokenKind::Set => {
                            let name = match self.next().unwrap_or(EOF.clone()).kind {
                                TokenKind::Identifier(name) => name,
                                _ => {
                                    return Err(ParseError {
                                        message: "expected identifier".to_string(),
                                    })
                                }
                            };

                            let params = self.parse_params()?;
                            assert_eq!(self.next().unwrap_or(EOF.clone()).kind, TokenKind::Colon);
                            let _ret = self.parse_type_ann()?;

                            // TODO: check that _ret is undefined

                            props.push(ObjectProp::Setter(ObjSetter {
                                name,
                                params,
                                span: Span { start: 0, end: 0 }, // TODO
                            }));
                        }
                        _ => {
                            return Err(ParseError {
                                message: "expected identifier or indexer".to_string(),
                            })
                        }
                    }

                    match self.peek().unwrap_or(&EOF).kind {
                        TokenKind::Comma => {
                            self.next();
                        }
                        TokenKind::RightBrace => {
                            break;
                        }
                        _ => {
                            return Err(ParseError {
                                message: "expected ',' or '}'".to_string(),
                            })
                        }
                    }
                }

                span = merge_spans(&span, &self.peek().unwrap_or(&EOF).span);
                assert_eq!(
                    self.next().unwrap_or(EOF.clone()).kind,
                    TokenKind::RightBrace
                );

                TypeAnnKind::Object(props)
            }
            TokenKind::LeftBracket => {
                self.next(); // consumes '['
                let mut elems: Vec<TypeAnn> = vec![];

                while self.peek().unwrap_or(&EOF).kind != TokenKind::RightBracket {
                    elems.push(self.parse_type_ann()?);

                    if self.peek().unwrap_or(&EOF).kind == TokenKind::Comma {
                        self.next().unwrap_or(EOF.clone());
                    } else {
                        break;
                    }
                }

                span = merge_spans(&span, &self.peek().unwrap_or(&EOF).span);
                assert_eq!(
                    self.next().unwrap_or(EOF.clone()).kind,
                    TokenKind::RightBracket
                );

                TypeAnnKind::Tuple(elems)
            }
            TokenKind::LeftParen => {
                let atom = self.parse_inside_parens(|p| p.parse_type_ann())?;
                return Ok(atom);
            }
            TokenKind::Identifier(ident) => {
                self.next(); // consumes identifier

                if self.peek().unwrap_or(&EOF).kind == TokenKind::LessThan {
                    self.next().unwrap_or(EOF.clone());
                    let mut params: Vec<TypeAnn> = vec![];

                    while self.peek().unwrap_or(&EOF).kind != TokenKind::GreaterThan {
                        params.push(self.parse_type_ann()?);

                        if self.peek().unwrap_or(&EOF).kind == TokenKind::Comma {
                            self.next().unwrap_or(EOF.clone());
                        } else {
                            break;
                        }
                    }

                    span = merge_spans(&span, &self.peek().unwrap_or(&EOF).span);
                    assert_eq!(
                        self.next().unwrap_or(EOF.clone()).kind,
                        TokenKind::GreaterThan
                    );

                    TypeAnnKind::TypeRef(ident, Some(params))
                } else {
                    TypeAnnKind::TypeRef(ident, None)
                }
            }
            TokenKind::Fn => {
                self.next(); // consumes 'fn'

                let type_params = self.maybe_parse_type_params()?;
                let params = self.parse_params()?;
                assert_eq!(self.next().unwrap_or(EOF.clone()).kind, TokenKind::Arrow);
                let return_type = self.parse_type_ann()?;

                // TODO: handle generics
                TypeAnnKind::Function(FunctionType {
                    type_params,
                    params,
                    ret: Box::new(return_type),
                })
            }
            TokenKind::KeyOf => {
                self.next(); // consumes 'keyof'

                let type_ann = self.parse_type_ann()?;

                TypeAnnKind::KeyOf(Box::new(type_ann))
            }
            TokenKind::Mut => {
                self.next(); // consumes 'mut'

                let type_ann = self.parse_type_ann()?;

                TypeAnnKind::Mutable(Box::new(type_ann))
            }
            TokenKind::TypeOf => {
                self.next(); // consumes 'typeof'

                let expr = self.parse_expr()?;

                TypeAnnKind::TypeOf(Box::new(expr))
            }
            token => {
                panic!("expected token to start type annotation, found {:?}", token)
            }
        };

        let atom = TypeAnn {
            kind,
            span,
            inferred_type: None,
        };

        Ok(atom)
    }

    fn parse_type_ann_postfix(
        &mut self,
        lhs: TypeAnn,
        next_precedence: (u8, Associativity),
    ) -> Result<TypeAnn, ParseError> {
        let _precedence = if next_precedence.1 == Associativity::Left {
            next_precedence.0
        } else {
            next_precedence.0 - 1
        };

        let token = self.peek().unwrap_or(&EOF).clone();

        let type_ann = match &token.kind {
            // TODO: handle parsing index access type
            TokenKind::LeftBracket => {
                self.next();
                match self.peek().unwrap_or(&EOF).kind {
                    TokenKind::RightBracket => {
                        let next = self.next().unwrap_or(EOF.clone());
                        let span = merge_spans(&lhs.span, &next.span);
                        TypeAnn {
                            kind: TypeAnnKind::Array(Box::new(lhs)),
                            span,
                            inferred_type: None,
                        }
                    }
                    _ => {
                        let index_type = self.parse_type_ann()?;
                        let merged_span = merge_spans(&lhs.span, &index_type.span);
                        assert_eq!(
                            self.next().unwrap_or(EOF.clone()).kind,
                            TokenKind::RightBracket
                        );
                        TypeAnn {
                            kind: TypeAnnKind::IndexedAccess(Box::new(lhs), Box::new(index_type)),
                            span: merged_span,
                            inferred_type: None,
                        }
                    }
                }
                // let next = self.peek().unwrap_or(&EOF);
                // let merged_span = merge_spans(&lhs.span, &next.span);
                // assert_eq!(
                //     self.next().unwrap_or(EOF.clone()).kind,
                //     TokenKind::RightBracket
                // );
                // TypeAnn {
                //     kind: TypeAnnKind::Array(Box::new(lhs)),
                //     span: merged_span,
                //     inferred_type: None,
                // }
            }
            _ => panic!("unexpected token: {:?}", token),
        };

        Ok(type_ann)
    }

    fn parse_type_ann_with_precedence(&mut self, precedence: u8) -> Result<TypeAnn, ParseError> {
        let mut lhs = self.parse_type_ann_atom()?;

        loop {
            let next = self.peek().unwrap_or(&EOF).clone();
            if let TokenKind::Eof = next.kind {
                return Ok(lhs);
            }

            if let TokenKind::Semicolon = next.kind {
                return Ok(lhs);
            }

            if let Some(next_precedence) = get_postfix_precedence(&next) {
                if precedence >= next_precedence.0 {
                    return Ok(lhs);
                }

                lhs = self.parse_type_ann_postfix(lhs.clone(), next_precedence)?;

                continue;
            }

            if let Some(next_precedence) = get_infix_precedence(&next) {
                if precedence >= next_precedence.0 {
                    return Ok(lhs);
                }

                self.next();

                let precedence = if next_precedence.1 == Associativity::Left {
                    next_precedence.0
                } else {
                    next_precedence.0 - 1
                };

                lhs = match &next.kind {
                    TokenKind::Ampersand => {
                        let start = lhs.span.start;
                        let rhs = self.parse_type_ann_with_precedence(precedence)?;
                        let mut end = rhs.span.end;
                        let mut types = vec![lhs, rhs];
                        while TokenKind::Ampersand == self.peek().unwrap_or(&EOF).kind {
                            self.next();
                            let rhs = self.parse_type_ann_with_precedence(precedence)?;
                            end = rhs.span.end;
                            types.push(rhs);
                        }
                        let span = Span { start, end };

                        TypeAnn {
                            kind: TypeAnnKind::Intersection(types),
                            span,
                            inferred_type: None,
                        }
                    }
                    TokenKind::Pipe => {
                        let start = lhs.span.start;
                        let rhs = self.parse_type_ann_with_precedence(precedence)?;
                        let mut end = rhs.span.end;
                        let mut types = vec![lhs, rhs];
                        while TokenKind::Pipe == self.peek().unwrap_or(&EOF).kind {
                            self.next();
                            let rhs = self.parse_type_ann_with_precedence(precedence)?;
                            end = rhs.span.end;
                            types.push(rhs);
                        }
                        let span = Span { start, end };

                        TypeAnn {
                            kind: TypeAnnKind::Union(types),
                            span,
                            inferred_type: None,
                        }
                    }
                    _ => panic!("unexpected token {:?}", next.kind),
                };

                continue;
            }

            return Ok(lhs);
        }
    }

    pub fn parse_type_ann(&mut self) -> Result<TypeAnn, ParseError> {
        self.parse_type_ann_with_precedence(0)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::parser::Parser;

    pub fn parse(input: &str) -> TypeAnn {
        let mut parser = Parser::new(input);
        parser.parse_type_ann().unwrap()
    }

    #[test]
    fn parse_literal_types() {
        insta::assert_debug_snapshot!(parse("123"));
        insta::assert_debug_snapshot!(parse("true"));
        insta::assert_debug_snapshot!(parse("false"));
        insta::assert_debug_snapshot!(parse("null"));
        insta::assert_debug_snapshot!(parse("undefined"));
        insta::assert_debug_snapshot!(parse(r#""hello""#));
    }

    #[test]
    fn parse_primitive_types() {
        insta::assert_debug_snapshot!(parse("number"));
        insta::assert_debug_snapshot!(parse("string"));
        insta::assert_debug_snapshot!(parse("boolean"));
        insta::assert_debug_snapshot!(parse("symbol"));
    }

    #[test]
    fn parse_object_types() {
        insta::assert_debug_snapshot!(parse("{a: number, b?: string, c: boolean}"));
        insta::assert_debug_snapshot!(parse("{a: {b: {c: boolean}}}"));
        insta::assert_debug_snapshot!(parse("{\n  a: number,\n  b?: string,\n  c: boolean,\n}"));
    }

    #[test]
    fn parse_object_type_all_sig_types() -> Result<(), ParseError> {
        let input = r#"
            {
                fn (a: number): string,
                fn foo(a: number): string,
                fn bar(self, a: number): string,
                get baz(): string,
                set baz(value: string): undefined,
                [key: string]: number,
                qux: string,
            }
        "#;
        let mut parser = Parser::new(input);
        let result = parser.parse_type_ann()?;
        insta::assert_debug_snapshot!(result);

        Ok(())
    }

    #[test]
    #[should_panic]
    fn parse_object_type_missing_comma() {
        insta::assert_debug_snapshot!(parse("{a: number b: string}"));
    }

    #[test]
    #[should_panic]
    fn parse_object_type_missing_right_brace() {
        insta::assert_debug_snapshot!(parse("{a: number, b: string"));
    }

    #[test]
    fn parse_tuple_types() {
        insta::assert_debug_snapshot!(parse("[number, string, boolean]"));
        insta::assert_debug_snapshot!(parse("[\n  number,\n  string,\n  boolean,\n]"));
    }

    #[test]
    #[should_panic]
    fn parse_tuple_type_missing_comma() {
        insta::assert_debug_snapshot!(parse("[number string]"));
    }

    #[test]
    #[should_panic]
    fn parse_tuple_type_missing_right_bracket() {
        insta::assert_debug_snapshot!(parse("[number, string"));
    }

    #[test]
    fn parse_array_types() {
        insta::assert_debug_snapshot!(parse("number[]"));
        insta::assert_debug_snapshot!(parse("{x: number, y: number}[]"));
        insta::assert_debug_snapshot!(parse("T[][]"));
    }

    #[test]
    fn parse_type_refs() {
        insta::assert_debug_snapshot!(parse("Array<T>"));
        insta::assert_debug_snapshot!(parse("Map<K, V>"));
        insta::assert_debug_snapshot!(parse("Array<Array<T>>"));
        insta::assert_debug_snapshot!(parse("T"));
    }

    #[test]
    fn parse_fn_type_ann() {
        insta::assert_debug_snapshot!(parse("fn (a: number, b: number) => number"));
    }

    #[test]
    fn parse_union_types() {
        insta::assert_debug_snapshot!(parse("number | string"));
        insta::assert_debug_snapshot!(parse("number | string | boolean"));
    }

    #[test]
    fn parse_intersection_types() {
        insta::assert_debug_snapshot!(parse("number & string"));
        insta::assert_debug_snapshot!(parse("number & string & boolean"));
    }

    #[test]
    fn parse_union_and_intersection_combo() {
        insta::assert_debug_snapshot!(parse("number | string & boolean"));
        insta::assert_debug_snapshot!(parse("number & string | boolean"));
    }

    #[test]
    fn parse_parens_for_grouping() {
        insta::assert_debug_snapshot!(parse("number & (string | boolean)"));
    }

    #[test]
    fn parse_indexed_access() {
        insta::assert_debug_snapshot!(parse("T[K]"));
        insta::assert_debug_snapshot!(parse(r#"T["foo"]"#));
    }

    #[test]
    fn parse_indexer_type() {
        insta::assert_debug_snapshot!(parse("{[key: string]: number}"));
        insta::assert_debug_snapshot!(parse("{[key: string]: number, [key: number]: string}"));
    }
}
