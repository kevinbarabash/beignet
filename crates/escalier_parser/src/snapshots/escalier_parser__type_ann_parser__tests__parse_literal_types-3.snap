---
source: crates/escalier_parser/src/type_ann_parser.rs
expression: "parse(\"false\")"
---
TypeAnn {
    kind: BoolLit(
        false,
    ),
    loc: SourceLocation {
        start: Position {
            line: 1,
            column: 1,
        },
        end: Position {
            line: 1,
            column: 6,
        },
    },
}
